File: algo/problems/0242-valid-anagram/scala/ValidAnagram.scala
object ValidAnagram:
  /** 用 groupMapReduce 统计两边字符频次再比较（Scala 2.13+/3 的现代写法）。O(n)/O(k)。 */
  def isAnagram(s: String, t: String): Boolean =
    if s.length != t.length then false
    else
      def freq(str: String): Map[Char, Int] =
        str.groupMapReduce(identity)(_ => 1)(_ + _)
      freq(s) == freq(t)

File: algo/problems/0242-valid-anagram/go/validanagram.go
package validanagram

// IsAnagram 用计数 map 统计 s，再用 t 抵消；任一字符计数变负即非异位词。O(n)/O(k)。
// 注意：len 比的是字节数，按 rune 遍历可正确处理多字节字符。
func IsAnagram(s, t string) bool {
	if len(s) != len(t) {
		return false
	}
	count := make(map[rune]int)
	for _, c := range s {
		count[c]++
	}
	for _, c := range t {
		count[c]--
		if count[c] < 0 {
			return false
		}
	}
	return true
}
